### ISSUE TITLE
AttributeError: 'NeuralNetClassifier' object has no attribute 'score'

### ISSUE THREAD

**Issue by damienlancry** [skorch-dev/skorch#469]
Hi

I am trying to do active active learning with deep neural networks. for this i am using modAL which is a python library designed for active learning purposes and compatible with sklearn estimators. 
Since skorch is a sklearn wrapper for pytorch i was expecting it to work with modAL but with thtis minimal code:
```
from keras.datasets import mnist
from sklearn.model_selection import train_test_split
from skorch import NeuralNetClassifier
from torch import nn, optim, tensor, cuda
import torch.nn.functional as F 
from modAL import ActiveLearner

class ClassifierModule(nn.Module):
    def __init__(self):
        super(ClassifierModule, self).__init__()
        self.layer1 = nn.Sequential(
        nn.Conv2d(1, 32, 3),
        nn.ReLU(),
        nn.Conv2d(32, 32, 3),
        nn.ReLU(),
        nn.MaxPool2d(2, 2))
        self.d1 = nn.Dropout(0.25)
        self.f1 = nn.Linear(12 * 12 * 32, 128)
        self.d2 = nn.Dropout(0.5)
        self.f2 = nn.Linear(128, 10)
    def forward(self, x):
        out = x
        out = self.layer1(out)
        out = out.view(-1,12 * 12 * 32)
        out = self.d1(out)
        out = self.f1(out)
        out = self.d2(out)
        out = self.f2(out)
        return F.softmax(out, dim=-1)


(X, y), (X_test, y_test) = mnist.load_data()
X, X_test = X.reshape((-1,1,28,28)) / 255., X_test.reshape((-1,1,28,28)) / 255.

X_train, X_pool, y_train, y_pool = train_test_split(X, y, train_size=1000, random_state=42)

device = "cuda" if cuda.is_available() else "cpu"

X_train, X_pool = tensor(X_train).float().to(device), tensor(X_pool).float().to(device)
y_train, y_pool = tensor(y_train).long().to(device), tensor(y_pool).long().to(device)

model   = NeuralNetClassifier(ClassifierModule, max_epochs=100, batch_size=128, lr=0.001, device=device,
                              optimizer=optim.Adam, verbose=1)
learner = ActiveLearner(
    estimator=model,
    X_training=X_train, 
    y_training=y_train
)

model_accuracy = learner.score(X_test,y_test)
performance_history = [model_accuracy]
for index in range(100):
    query_index, query_instance = learner.query(X_pool)
    x, y = X_pool[query_index], y_pool[query_index]
    learner.teach(x,y)
    X_pool, y_pool = np.delete(X_pool, query_index, axis=0), np.delete(y_pool, query_index, axis=0)
    model_accuracy = learner.score(X_test, y_test)
    performance_history.append(model_accuracy)

```
I get the following error:
```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-17-91e0c53d9ed2> in <module>
     48 )
     49 
---> 50 model_accuracy = learner.score(X_test,y_test)
     51 performance_history = [model_accuracy]
     52 for index in range(100):

~/anaconda3/lib/python3.7/site-packages/modAL/models/base.py in score(self, X, y, **score_kwargs)
    207             The score of the predictor.
    208         """
--> 209         return self.estimator.score(X, y, **score_kwargs)
    210 
    211     @abc.abstractmethod

AttributeError: 'NeuralNetClassifier' object has no attribute 'score'
```
So i was wondering: Would not it be good to have a score method for NeuralNetClassifier objects?

Best
Damien

**Comment by ottonemo** [comment]
For the more specific classes such as `NeuralNetClassifier` and `NeuralNetRegressor` it might be possible to define default scoring methods but I am not sure if we will find an implementation that will fit most users. If you really want a `score` method, you can always subclass your `NeuralNet*` class and implement it yourself.

As I understand from the [modAL documentation](https://modal-python.readthedocs.io/en/latest/content/models/ActiveLearner.html#prediction-and-scoring) it is not strictly necessary to have a `score` method and you can simply use any sklearn metric on the model. Wouldn't that work for you?

**Comment by damienlancry** [comment]
yes that works! Thanks a lot. But I still think that a score method would be great! just like KerasClassifier and KerasRegressor objects have.

**Comment by H4dr1en** [comment]
I just encountered the same error with the following small reproducible code:

```
import numpy as np
from sklearn.datasets import make_regression
from skorch import NeuralNetRegressor
from skorch.toy import make_regressor
from sklearn.model_selection import train_test_split

X, Y = make_regression(n_samples=int(5e4), n_features=50, n_informative=25, bias=-92, noise=100)
X = X.astype(np.float32)
Y = Y.astype(np.float32)
X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.2, random_state=42)

regressor = make_regressor(input_units=50, hidden_units=100, dropout=0.4)()

estimator = NeuralNetRegressor(regressor,
    max_epochs=10,
    lr=1e-3,    
    iterator_train__shuffle=True,# Shuffle training data on each epoch
)
estimator.score(X_test, y_test )
```

Since the aim of skorch is to have Pytorch NNs being sklearn compatible, one would expect `NeuralNetRegressor` is [RegressorMixin](https://scikit-learn.org/stable/modules/generated/sklearn.base.RegressorMixin.html#sklearn.base.RegressorMixin) compatible and therefore implement `score`, right?

**Comment by thomasjpfan** [comment]
It looks like `Keras` follows the scikit-learn convention where `score` is accuracy for classifiers and r2_score for regressors. It would not be too difficult to implement this in `skorch`.

**Comment by BenjaminBossan** [comment]
I think I could live with accuracy, though this metric is also often problematic. But R² just makes so little sense, I don't like it at all. The pro of having `score` would be compatibility, the pro of not having `score` is that it makes users think about what they want. Maybe a compromise is to have `score` but to raise a helpful error message when it is called?

**Comment by thomasjpfan** [comment]
All scikit-learn estimators have a `score` method, since they inherit from `RegressorMixin` or `ClassifierMixin`. Inside of these mixins, they add a `score` method that internally calls `r2_score` or `accuracy_score` with `self.predict`.

**Comment by damienlancry** [comment]
I opened a PR #470 . For my purposes it works very well.

**Comment by BenjaminBossan** [comment]
Was solved in #470
